src/frontend/src/binder/relation/mod.rs

```diff
@@ -377,16 +378,23 @@ impl Binder {
                     Ok(Relation::BackCteRef(Box::new(BoundBackCteRef { share_id })))
                 }
                 BindingCteState::Bound { query } => {
+                    let schema = match query.clone() {
+                        Left(normal) => normal.body.schema().clone(),
+                        Right(recursive) => recursive.recursive.body.schema().clone(),
+                    };
                     self.bind_table_to_context(
-                        query
-                            .body
-                            .schema()
+                        schema
                             .fields
                             .iter()
                             .map(|f| (false, f.clone())),
                         table_name.clone(),
                         Some(original_alias),
                     )?;
+                    // todo: to be further reviewed
+                    let query = match query {
+                        Left(normal) => normal,
+                        Right(recursive) => *recursive.recursive,
```

Comment:
current choice is to choose the bound `recursive` as the query to share, if the cte is recursive.
I'm wondering is this semantically correct in all cases? 🤔

Reply:
No `*recursive.recursive` is just a part of the recursive CTE. We should bind `RecursiveUnion` as a whole here.